Single-key matching behavior in generic.Matcher

lavalamp opened the pull request:
People should be able to get the single-key-watching behavior without using the generic function if they wish. I removed WatchSingleFieldName from generic.Etcd, I think this way is clearer: if we want this behavior by default, I think generic's SelectionPredicate is a better home for it.

@wojtek-t 

Note that I remove a test in minion/etcd, and replace with two tests: one in minion/rest_test.go that verifies that the matcher is correct, and one in generic/etcd that verifies that watching on a single key works.

lavalamp committed:
Single-key matching behavior in generic.Matcher

smarterclayton commented on pkg/registry/minion/rest_test.go:
```diff
@@ -0,0 +1,45 @@
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+)
+
+func TestMatchNode(t *testing.T) {
```
Rather than test here you could add this to resttest as TestWatch and verify singular behavior.  Means you can add it to any supported storage.

lavalamp replied:
Sorry--I don't follow--is there a fancy new testing package I'm not aware of?

smarterclayton replied:
Yeah, for testing that storage objects respect the rest conventions.  It's a bit higher level than this, but see pkg/api/rest/resttest.  


> On Apr 2, 2015, at 4:36 PM, Daniel Smith <notifications@github.com> wrote:
> 
> In pkg/registry/minion/rest_test.go:
> 
> > +distributed under the License is distributed on an "AS IS" BASIS,
> > +WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
> > +See the License for the specific language governing permissions and
> > +limitations under the License.
> > +*/
> > +
> > +package minion
> > +
> > +import (
> > +	"testing"
> > +
> > +	"github.com/GoogleCloudPlatform/kubernetes/pkg/fields"
> > +	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
> > +)
> > +
> > +func TestMatchNode(t *testing.T) {
> Sorry--I don't follow--is there a fancy new testing package I'm not aware of?
> 
> —
> Reply to this email directly or view it on GitHub.
>

lavalamp replied:
OK I will give that a shot, that looks like a useful thing.

smarterclayton commented:
LGTM - one comment about expanding resttest

wojtek-t commented on pkg/registry/generic/etcd/etcd.go:
```diff
@@ -407,30 +403,30 @@ func (e *Etcd) finalizeDelete(obj runtime.Object, runHooks bool) (runtime.Object
+		if err != nil {
+			return nil, err
+		}
+		// This matches a single object-- no need to call the predicate all the time.
```
I'm wondering if that's always true. It's technically possible that you specify to match a single object that additionally matches some labels. Don't we want to support that case too? (i.e. don't we want to call this predicate anyway even if we are watching a single object?)

lavalamp replied:
hm... what would the use case be for that? Like you want to monitor a pod and get notifications only when it has particular labels applied? I was thinking that if you try to do both it's probably a user error. I guess I could change it.

lavalamp replied:
OK, I think you're right-- I think I actually broke list with this change. Troubling that no test caught it.

lavalamp replied:
I also realized that I need to be much more explicit about the handling of namespace in the comments.

wojtek-t commented on pkg/registry/generic/etcd/etcd_test.go:
```diff
@@ -97,13 +97,21 @@ func (sm SetMatcher) Matches(obj runtime.Object) (bool, error) {
-// EverythingMatcher matches everything
-type EverythingMatcher struct{}
+func (setMatcher) MatchesSingle() (string, bool) {
+	return "", false
```
I think it should return true if the size of StringSet is set to 1.

smarterclayton replied:
Why would that be guaranteed to match a single resource?  Field "portalIP=None" would have a single selector but match many services.


> On Apr 3, 2015, at 2:11 AM, Wojciech Tyczynski <notifications@github.com> wrote:
> 
> In pkg/registry/generic/etcd/etcd_test.go:
> 
> > @@ -97,13 +97,21 @@ func (sm SetMatcher) Matches(obj runtime.Object) (bool, error) {
> >  	return sm.Has(pod.Name), nil
> >  }
> >  
> > -// EverythingMatcher matches everything
> > -type EverythingMatcher struct{}
> > +func (setMatcher) MatchesSingle() (string, bool) {
> > +	return "", false
> I think it should return true if the size of StringSet is set to 1.
> 
> —
> Reply to this email directly or view it on GitHub.
>

wojtek-t replied:
IIUC setMatcher contains only the list of names that we would like to match (at least it's how I understood it from Matches method and its line 97)

smarterclayton replied:
My bad, you're right.


> On Apr 3, 2015, at 11:10 AM, Wojciech Tyczynski <notifications@github.com> wrote:
> 
> In pkg/registry/generic/etcd/etcd_test.go:
> 
> > @@ -97,13 +97,21 @@ func (sm SetMatcher) Matches(obj runtime.Object) (bool, error) {
> >  	return sm.Has(pod.Name), nil
> >  }
> >  
> > -// EverythingMatcher matches everything
> > -type EverythingMatcher struct{}
> > +func (setMatcher) MatchesSingle() (string, bool) {
> > +	return "", false
> IIUC setMatcher contains only the list of names that we would like to match (at least it's how I understood it from Matches method and its line 97)
> 
> —
> Reply to this email directly or view it on GitHub.
>

lavalamp replied:
Oh, I think you're right-- it matches by name which is indeed its key. Good catch.

wojtek-t commented:
LGTM - thanks for improving my changes.

lavalamp commented on pkg/registry/minion/rest.go:
```diff
@@ -91,6 +91,9 @@ func NodeToSelectableFields(node *api.Node) labels.Set {
 
 // MatchNode returns a generic matcher for a given label and field selector.
 func MatchNode(label labels.Selector, field fields.Selector) generic.Matcher {
+	if name, ok := field.RequiresExactMatch("name"); ok {
```
Hmmmm... shouldn't this be metadata.name, not just name?

wojtek-t replied:
I think it should be name. This is related to how fields are translated into lables in the conversion code, e.g.:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/v1beta3/conversion.go#L25
For all: ReplicationController, Pod and Node we have "name". 
But I agree with you that translating it to "metadata.name" would probably be better.

lavalamp replied:
I see, you added this in your change. Yes, just name by itself is incorrect. The field labels should match the access you would do to get the field in the object. "name" in v1beta3 is stored in .metadata.name. I will fix. This should have been caught in review, it's an api error.

wojtek-t replied:
Thanks. SGTM

lavalamp commented:
Updated. I moved this functionality into generic.SelectionPredicate, so *everything* should automagically get this fix. (because everything SHOULD be using the generic SelectionPredicate!)

@smarterclayton It seems like resttest package doesn't actually have a way to hook into etcd to verify the scope of the watches, so I'd like to defer adding this test there.

smarterclayton commented:
That's fine - I would expect resttest to accept an interface that validates that the underlying semantic operation completed, but haven't done many examples of that yet.


> On Apr 3, 2015, at 4:20 PM, Daniel Smith <notifications@github.com> wrote:
> 
> Updated. I moved this functionality into generic.SelectionPredicate, so everything should automagically get this fix. (because everything SHOULD be using the generic SelectionPredicate!)
> 
> @smarterclayton It seems like resttest package doesn't actually have a way to hook into etcd to verify the scope of the watches, so I'd like to defer adding this test there.
> 
> —
> Reply to this email directly or view it on GitHub.
>

lavalamp commented:
OK, thanks-- PTAL :)

wojtek-t commented:
Thanks @lavalamp - LGTM
